hazelcast/src/test/java/com/hazelcast/client/impl/connection/tcp/TcpClientConnectionManagerTranslateTest.java

```diff
@@ -72,9 +71,124 @@ public void teardown() {
+        // public member address is reachable
+        member.getAddressMap().put(EndpointQualifier.resolve(ProtocolType.CLIENT, "public"),
+                new Address("127.0.0.1", 5701));
+
+        // when
+        Connection connection = clientConnectionManager.getOrConnectToMember(member);
+
+        // then
+        assertNotNull(connection);
+    }
+
+    @Test(expected = Exception.class)
+    public void testTranslateIsNotUsedWhenPublicIpDisabled() throws UnknownHostException {
+        // given
+        ClientConfig clientConfig = new ClientConfig();
+        clientConfig.setProperty(ClientProperty.DISCOVERY_SPI_PUBLIC_IP_ENABLED.getName(), "false");
+
+        HazelcastInstance client = HazelcastClientUtil.newHazelcastClient(null, clientConfig);
+        TcpClientConnectionManager clientConnectionManager =
+                new TcpClientConnectionManager(getHazelcastClientInstanceImpl(client));
+        clientConnectionManager.start();
+
+        // private member address is incorrect
+        Member member = new MemberImpl(new Address("192.168.0.1", 5701), VERSION, false);
+        // public member address is correct
+        member.getAddressMap().put(EndpointQualifier.resolve(ProtocolType.CLIENT, "public"), new Address("127.0.0.1", 5701));
+
+        // when
+        clientConnectionManager.getOrConnectToMember(member);
+
+        // then
+        // throws exception because it can't connect to the incorrect member address
+    }
+
+    @Test(expected = Exception.class)
+    public void testTranslateFromMemberIsNotUsedWhenAlreadyTranslatedByAddressProvider() throws UnknownHostException {
+        // given
+        ClientConfig clientConfig = new ClientConfig();
+        clientConfig.setProperty(ClientProperty.DISCOVERY_SPI_PUBLIC_IP_ENABLED.getName(), "true");
+
+        TestAddressProvider provider = new TestAddressProvider(false);
+        HazelcastInstance client = HazelcastClientUtil.newHazelcastClient(provider, clientConfig);
+        TcpClientConnectionManager clientConnectionManager =
+                new TcpClientConnectionManager(getHazelcastClientInstanceImpl(client));
+        clientConnectionManager.start();
+        provider.shouldTranslate = true;
+        privateAddress = new Address("192.168.0.1", 5702);
+
+        // private member address is correct
+        Member member = new MemberImpl(privateAddress, VERSION, false);
+        // public member address is correct
+        member.getAddressMap().put(EndpointQualifier.resolve(ProtocolType.CLIENT, "public"),
+                new Address("127.0.0.1", 5701));
+
+        // when
+        clientConnectionManager.getOrConnectToMember(member);
+
+        // then
+        // throws exception because it can't connect to the incorrect address
+    }
+
     private class TestAddressProvider implements AddressProvider {
 
-        volatile boolean shouldTranslate = false;
+        private boolean shouldTranslate;
```

Comment:
Is it ok to remove `volatile` modifier for this field?

Reply:
It should not cause any harm as far as I see. We set this field before starting the instance. All threads start after this.